src/Lykke.Service.ExchangeConnector/Exchanges/Concrete/Shared/OrderBooksHarvesterBase.cs

```diff
@@ -1,46 +1,78 @@
 ﻿using System;
+using System.Collections.Concurrent;
 using System.Collections.Generic;
 using System.Linq;
 using System.Threading;
 using System.Threading.Tasks;
 using Common.Log;
 using Polly;
-using TradingBot.Exchanges.Concrete.BitMEX;
+using TradingBot.Communications;
 using TradingBot.Infrastructure.Configuration;
+using TradingBot.Infrastructure.Exceptions;
 using TradingBot.Trading;
 
 namespace TradingBot.Exchanges.Concrete.Shared
 {
     internal abstract class OrderBooksHarvesterBase : IDisposable
     {
         protected readonly ILog Log;
-        protected readonly WebSocketTextMessenger Messenger;
-        protected readonly ISet<OrderBookItem> OrderBookSnapshot;
+        protected readonly OrderBookSnapshotsRepository OrderBookSnapshotsRepository;
+        protected readonly OrderBookEventsRepository OrderBookEventsRepository;
+        protected CancellationToken CancellationToken;
+
+        private readonly ConcurrentDictionary<string, OrderBookSnapshot> _orderBookSnapshots;
+        private ExchangeConverters _converters;
+        private readonly Timer _heartBeatMonitoringTimer;
+        private readonly TimeSpan _heartBeatPeriod = TimeSpan.FromSeconds(30);
+        private CancellationTokenSource _cancellationTokenSource;
         private Task _messageLoopTask;
         private Func<OrderBook, Task> _newOrderBookHandler;
-        protected ICurrencyMappingProvider CurrencyMappingProvider { get; }
-        private CancellationTokenSource _cancellationTokenSource;
-        protected CancellationToken CancellationToken;
         private DateTime _lastPublishTime = DateTime.MinValue;
         private long _lastSecPublicationsNum;
         private int _orderBooksReceivedInLastTimeFrame;
-        private readonly Timer _heartBeatMonitoringTimer;
-        private readonly TimeSpan _heartBeatPeriod = TimeSpan.FromSeconds(30);
         private Task _measureTask;
         private long _publishedToRabbit;
 
-        protected OrderBooksHarvesterBase(ICurrencyMappingProvider currencyMappingProvider, string uri, ILog log)
+        protected IExchangeConfiguration ExchangeConfiguration { get; }
+
+        private string _exchangeName;
+        public string ExchangeName
+        {
+            get => _exchangeName;
+            set
+            {
+                _exchangeName = value;
+                _converters = new ExchangeConverters(
```

Comment:
Why do we need to recreate _converters on exchange name setting?

Reply:
Fixed

Reply:
Converters are immutable class and some of the methods rely on the exchange name property.

Reply:
So let's pass exchange name as a constructor argument into OrderBookHarvesterBase and create _converters from there with correct name from the beginning.

Reply:
OrderBookHarvesterBase inherited instances are actually injected automatically from the IoC, so that's why the name is set later.